### PR TITLE
Allocate memory using lanl/SICM

### PR DESCRIPTION
**Pull request by calccrypto**
Added SICM resource, allocator, and strategy
Added SICM cookbook recipe
Added script to install SICM on Travis CI
Added sphinx documentation
Updated changelog

### CODE REVIEW

File: src/umpire/alloc/SICMAllocator.hpp [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    {
+      std::lock_guard <std::mutex> lock(arena_mutex);
+
+      if (!allowed_devices.size()) {
+        cleanup();
+        UMPIRE_ERROR("SICM( bytes = " << bytes << " ) failed due to lack of allowed devics");
+      }
+
+      // find best device
+      const int best = allowed_devices[best_device_index];
```
Comment (calccrypto) [inline src/umpire/alloc/SICMAllocator.hpp]: This needs to be changed to not just round robin the devices.

File: src/umpire/alloc/SICMAllocator.hpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        if (!(sa = sicm_arena_create(0, &devs.devices[best]))) {
+          cleanup();
+          UMPIRE_ERROR("SICMAllocator Could not create arena on device " << best);
+        }
+      }
+      else {
+        UMPIRE_LOG(Debug, "Using existing arena on device " << best);
+
+        // take first arena
+        sa = arenas_on_device.front();
```
Comment (calccrypto) [inline src/umpire/alloc/SICMAllocator.hpp]: Is round robin-ing the arenas a good idea?

File: src/umpire/ResourceManager.cpp [hunk 3]
```diff
@@ -77,10 +82,30 @@ ResourceManager::ResourceManager() :
   resource::MemoryResourceRegistry& registry =
     resource::MemoryResourceRegistry::getInstance();
 
+// read config file/figure out which device belongs to what
+
+#if defined(UMPIRE_ENABLE_SICM)
+  registry.registerMemoryResource(
+      new resource::SICMResourceFactory("HOST", {0, 1, 2, 3, 4, 5}));
```
Comment (calccrypto) [inline src/umpire/ResourceManager.cpp]: The device list should eventually be obtained from a function.

File: src/umpire/alloc/CMakeLists.txt [hunk 4]
```diff
@@ -35,23 +45,38 @@ if (ENABLE_HCC)
...
-  $<BUILD_INTERFACE:${CMAKE_BINARY_DIR}/include>
-  $<INSTALL_INTERFACE:include>)
-
-install(TARGETS
-  umpire_alloc
-  EXPORT umpire-targets
-  RUNTIME DESTINATION lib
-  LIBRARY DESTINATION lib
-  ARCHIVE DESTINATION lib)
+if ("${umpire_alloc_sources}" STREQUAL "")
```
Comment (calccrypto) [inline src/umpire/alloc/CMakeLists.txt]: These changes to the cmake might not be correct.

File: src/umpire/ResourceManager.cpp [hunk 5]
```diff
@@ -77,10 +83,39 @@ ResourceManager::ResourceManager() :
   resource::MemoryResourceRegistry& registry =
     resource::MemoryResourceRegistry::getInstance();
 
+#if defined(UMPIRE_ENABLE_SICM)
+  // read config file/figure out which device belongs to what
```
Comment (calccrypto) [inline src/umpire/ResourceManager.cpp]: An actual implementation is needed here.

File: src/umpire/ResourceManager.cpp [hunk 6]
```diff
@@ -77,10 +83,39 @@ ResourceManager::ResourceManager() :
...
+  const std::vector <int> nodes = numa::get_allocatable_nodes();
```
Comment (jdahm) [inline src/umpire/ResourceManager.cpp]: FYI I believe using these functions requires `ENABLE_NUMA=ON`.

Comment (calccrypto) [reply src/umpire/ResourceManager.cpp]: I modified the cmake stuff to build some of the numa stuff when `ENABLE_SICM=ON`, even though `ENABLE_NUMA=OFF`

File: src/umpire/ResourceManager.cpp [hunk 7]
```diff
@@ -93,13 +128,22 @@ ResourceManager::ResourceManager() :
   registry.registerMemoryResource(
     new resource::CudaConstantMemoryResourceFactory());
 #endif
+#endif
 
 #if defined(UMPIRE_ENABLE_HCC)
+#if defined(UMPIRE_ENABLE_SICM)
```
Comment (davidbeckingsale) [inline src/umpire/ResourceManager.cpp]: This might be a bit premature.

Comment (calccrypto) [reply src/umpire/ResourceManager.cpp]: Removed.

File: src/umpire/alloc/SICMAllocator.hpp [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+#include <cstdlib>
+#include <list>
+#include <map>
+#include <sched.h>
+#include <vector>
+
+#include "umpire/util/Macros.hpp"
+#include "umpire/util/SICM_device.hpp"
+
+extern "C"
```
Comment (jdahm) [inline src/umpire/alloc/SICMAllocator.hpp]: Have you considered adding this to SICM directly?
```
#ifdef __cplusplus
extern "C" {
#endif
// ...
#ifdef __cplusplus
}
#endif
```

Comment (calccrypto) [reply src/umpire/alloc/SICMAllocator.hpp]: I've made a pull request in SICM. I'll hold off changing the `#include`s here until the pull request has been merged.

Comment (calccrypto) [reply src/umpire/alloc/SICMAllocator.hpp]: 7c23a51849e8f2fd669786a39444138807cd154b

File: src/umpire/alloc/CMakeLists.txt (hunk 4 above)
Comment (calccrypto) [reply src/umpire/alloc/CMakeLists.txt]: Reverted in 8c7aa413cc094ddff7aa0c594d0cd54327d4c67b

**Pull request closed without merging**